File: perl6/sample/church.pl
#!/usr/bin/env perl6

use v6;

my $zero = ->$f{->$x{$x}};
my $addone = ->$n{->$f{->$x{$f($n($f)($x))}}};
my $one = $addone($zero);
my $add = ->$m,$n{->$f{->$x{$n($f)($m($f)($x))}}};
my $two = $add($one, $one);
my $three = $add($two, $one);

my $inc = {$^n+1};
say 'zero  = ', $zero($inc)(0);
say 'one   = ', $one($inc)(0);
say 'two   = ', $two($inc)(0);
say 'three = ', $three($inc)(0);

File: ruby/sample/church.rb
#!/usr/bin/env ruby19
zero = lambda{|f|lambda{|x|x}}
addone = lambda{|m|lambda{|f|lambda{|x|f.call(m.call(f).call(x))}}}
one = addone.call(zero)
plus = lambda{|m,n|lambda{|f|lambda{|x|m.call(f).call(n.call(f).call(x))}}}
two = plus.call(one, one)
three = plus.call(one, two)

inc = lambda{|x|x+1}
puts 'zero  = ' + zero.call(inc).call(0).to_s
puts 'one   = ' + one.call(inc).call(0).to_s
puts 'two   = ' + two.call(inc).call(0).to_s
puts 'three = ' + three.call(inc).call(0).to_s
